### DISCUSSION TITLE
Integrated 24-bit and 32-bit pixel interface?

### DISCUSSION THREAD

**Discussion by Hakoyu** [SixLabors/ImageSharp#2472] [Ideas] [answered]
Like this
```Csharp
interface IPixel24
{
    public byte R;
    public byte G;
    public byte B;
}
interface IPixel32
{
    public byte A;
    public byte R;
    public byte G;
    public byte B;
}
```
`Bgr24`, `Rgb24` Inheritance `IPixel24`
`Abgr32`, `Argb32`, `Rgba32`, `Bgra32` Inheritance `IPixel32`
This operation makes it easier to get the RGB or ARGB value of a pixel without having to convert it for each pixel.
Because `IPixel<T>` does not contain RGB values directly, it is limited to use this as a generic function.

**Comment by JimBobSquarePants** [comment]
What's your use case that would require such functionality? Do you have an example of what you are trying to do?

**Reply by Hakoyu** [reply]
> What's your use case that would require such functionality? Do you have an example of what you are trying to do?

I want compare images using pixels, os I must know the RGB value of the pixel.
Because IPixel does not contain RGB values directly, providing support for every PixelFormats is quite tedious.

**Reply by Hakoyu** [reply]
Again, this can improve the reusability of some methods
`Argb32.FromPixel32(IPixel32)` can replace `FromBgra32`, `FromAbgr32` etc.

**Reply by JimBobSquarePants** [reply]
Why not decode each image to the specific Image<TPixel> you want and compare that?

**Reply by Hakoyu** [reply]
I have implemented this, but decoding reduces performance a bit.

**Reply by Hakoyu** [reply]
You know, no matter how efficient the conversion is written, the fastest way is still to process the raw data.

![image](https://github.com/SixLabors/ImageSharp/assets/48956359/5083c600-791d-47b0-87a2-14f17ec1555f)

So you can add a method to get the original format of the image, so that I do not have to write it myself.

![image](https://github.com/SixLabors/ImageSharp/assets/48956359/956bb64b-2f83-45e0-a7d5-b9848eaa8eeb)

**Reply by JimBobSquarePants** [reply, marked as answer]
That’s a terrible benchmark not comparing what I mean at all. 

use `Image.Load<Rgba32>` for both images. No need to clone anything.

**Reply by Hakoyu** [reply]
Well, I found that regardless of the generic type of `Image.Load<T>()`, it takes roughly the same amount of time.
Perhaps using only one pixel format would be a good enough solution.

**Reply by antonfirsov** [reply]
```C#
interface IPixel32
{
    public byte A { get; }
    public byte R { get; }
    public byte G { get; }
    public byte B { get; }
}
```

This would lead to 4 interface calls with virtual dispatch *for each pixel*, which would perform terribly. Our internal conversion algorithms are heavily optimized, so consolidating to a single pixel format should be fast.
